### PR TITLE
Collect and store LogDepositWithTokenId  and LogTokenRegistered events

### PR DESCRIPTION
**Pull request by Wendrowiec13**
Resolves L2B-619
Resolves L2B-620
Resolves L2B-621

Note to the reviewers:
Please be vary of the TODO comments. These were issues that I was not sure how to tackle. Suggestions would be appreciated

### CODE REVIEW

File: packages/backend/src/peripherals/ethereum/EthereumClient.ts [hunk 1]
```diff
@@ -162,6 +162,14 @@ export class EthereumClient {
     return await this.provider.getTransactionReceipt(transactionHash.toString())
   }
 
+  getContract(address: string, abi: string[]) {
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/ethereum/EthereumClient.ts:165]: It's unused now, remove it

File: packages/backend/src/peripherals/database/TokenRepository.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  tokenId: string | null
+  uri: string | null
+  contractError: unknown[]
+}
+
+export class TokenRepository extends BaseRepository {
+  constructor(database: Database, logger: Logger) {
+    super(database, logger)
+
+    /* eslint-disable @typescript-eslint/unbound-method */
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/TokenRepository.ts:21]: reenable at the end

File: packages/backend/src/peripherals/database/TokenRegistrationRepository.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  async addMany(records: TokenRegistrationRecord[]) {
+    const knex = await this.knex()
+    const rows = records.map(toRow)
+    const hashes = await knex('token_registrations')
+      .insert(rows)
+      .returning('asset_type_hash')
+    return hashes.map((x) => x.asset_type_hash)
+  }
+
+  async getAll(): Promise<TokenRegistrationRecord[]> {
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/TokenRegistrationRepository.ts:40]: not wrapped

File: packages/backend/src/peripherals/database/TokenRegistrationRepository.test.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  afterEach(() => repository.deleteAll())
+
+  it('adds single record and queries it', async () => {
+    const record: TokenRegistrationRecord = dummyTokenRegistration('1')
+
+    await repository.addMany([record])
+
+    const actual = await repository.getAll()
+
+    expect(actual).toEqual([{ ...record, contractError: [{}] }])
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/TokenRegistrationRepository.test.ts:25]: why is the contract error an array with a single empty object? Why isn't it an empty array?

**Review by sz-piotr — CHANGES_REQUESTED** [review]
I like it overall, but I'd love a quick pair progamming session to refactor this a little.

File: packages/backend/src/peripherals/database/TokenRegistrationRepository.test.ts (hunk 4 above)
Comment (Wendrowiec13) [reply packages/backend/src/peripherals/database/TokenRegistrationRepository.test.ts:25]: That's how an empty array is represented in the database using `json` type

File: packages/types/src/SpotAssetId.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+export interface SpotAssetId extends String {
+  _SpotAssetIdBrand: string
+}
+
+export function SpotAssetId(value: string) {
```
[the remark this answers is not in the document]
Comment (Wendrowiec13) [reply packages/types/src/SpotAssetId.ts:5]: So we could just use Hash256, right? Or do we want to somehow differentiate it from Hash256?

File: packages/backend/src/peripherals/database/shared/types.ts [hunk 6]
```diff
@@ -186,6 +186,25 @@ declare module 'knex/types/tables' {
...
+    name: string | null
+    symbol: string | null
+    quantum: string
+    decimals: number | null
+    contract_error: json
+  }
+
+  interface TokenRow {
+    asset_type_hash: string
+    // name: TODO: string Figure out how to get name for EIC-1155 standard
```
[the remark this answers is not in the document]
Comment (Wendrowiec13) [reply packages/backend/src/peripherals/database/shared/types.ts:202]: Yes, changed

File: packages/backend/src/peripherals/database/TokenRepository.ts (hunk 2 above)
Comment (Wendrowiec13) [reply packages/backend/src/peripherals/database/TokenRepository.ts:21]: I think all the other repositories also disable this rule. Do we also want to reenable those?

File: packages/backend/src/peripherals/database/TokenRegistrationRepository.test.ts (hunk 4 above)
Comment (sz-piotr) [reply packages/backend/src/peripherals/database/TokenRegistrationRepository.test.ts:25]: We need to do JSON.stringify when writing to this column: https://knexjs.org/guide/schema-builder.html#json

**Review by sz-piotr — APPROVED** [review]

**Pull request merged**